storage_client.go

```diff
@@ -13,9 +23,138 @@ type storageUtil interface {
+func (rsu *remoteStorageUtil) createClient(info *execResponseStageInfo, useAccelerateEndpoint bool) cloudClient {
+	utilClass := rsu.getNativeCloudType(info.LocationType).(cloudUtil)
+	return utilClass.createClient(info, useAccelerateEndpoint)
+}
+
+func (rsu *remoteStorageUtil) uploadOneFile(meta *fileMetadata) error {
+	var encryptMeta *encryptMetadata
+	var dataFile string
+	var err error
+	if meta.encryptionMaterial != nil {
+		if meta.srcStream != nil {
+			var encryptedStream bytes.Buffer
+			srcStream := meta.srcStream
+			if meta.realSrcStream != nil {
+				srcStream = meta.realSrcStream
+			}
+			encryptMeta, err = encryptStream(meta.encryptionMaterial, srcStream, &encryptedStream, 0)
+			if err != nil {
+				return err
+			}
+			meta.realSrcStream = &encryptedStream
+			dataFile = meta.realSrcFileName
+		} else {
+			encryptMeta, dataFile, err = encryptFile(meta.encryptionMaterial, meta.realSrcFileName, 0, meta.tmpDir)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		dataFile = meta.realSrcFileName
+	}
+
+	utilClass := rsu.getNativeCloudType(meta.stageInfo.LocationType)
+	maxConcurrency := int(meta.parallel)
+	var lastErr error
+	maxRetry := defaultMaxRetry
+	for retry := 0; retry < maxRetry; retry++ {
+		if !meta.overwrite {
+			header := utilClass.getFileHeader(meta, meta.dstFileName)
+			if header != nil && meta.resStatus == uploaded {
+				meta.dstFileSize = 0
+				meta.resStatus = skipped
+				return nil
+			}
+		}
+		if meta.overwrite || meta.resStatus == notFoundFile {
+			utilClass.uploadFile(dataFile, meta, encryptMeta, maxConcurrency, meta.multiPartThreshold)
+		}
+		if meta.resStatus == uploaded || meta.resStatus == renewToken || meta.resStatus == renewPresignedURL {
+			return nil
+		} else if meta.resStatus == needRetry {
+			lastErr = meta.lastError
+			if !meta.noSleepingTime {
+				sleepingTime := intMin(int(math.Exp2(float64(retry))), 16)
+				time.Sleep(time.Second * time.Duration(sleepingTime))
+			}
+		} else if meta.resStatus == needRetryWithLowerConcurrency {
+			lastErr = meta.lastError
+			maxConcurrency = int(meta.parallel) - (retry * int(meta.parallel) / maxRetry)
+			maxConcurrency = intMax(defaultConcurrency, maxConcurrency)
+			meta.lastMaxConcurrency = maxConcurrency
+
+			if !meta.noSleepingTime {
+				sleepingTime := intMin(int(math.Exp2(float64(retry))), 16)
+				time.Sleep(time.Second * time.Duration(sleepingTime))
+			}
+		}
+	}
+	if lastErr != nil {
+		return lastErr
+	}
+	return fmt.Errorf("unkown error uploading %v", dataFile)
+}
+
+func (rsu *remoteStorageUtil) uploadOneFileWithRetry(meta *fileMetadata) error {
+	utilClass := rsu.getNativeCloudType(meta.stageInfo.LocationType)
+	var retryFailed bool
+	for i := 0; i < 10; i++ {
+		// retry
+		err := rsu.uploadOneFile(meta)
+		if err != nil {
+			return err
+		}
+		if meta.resStatus == uploaded {
+			for j := 0; j < 10; j++ {
+				utilClass.getFileHeader(meta, meta.dstFileName)
+				if meta.resStatus == notFoundFile {
```

Comment:
Is there a test case covering this 'if meta.resStatus == notFoundFile' code path? If getFileHeader is done successfully then line 139 is always executed then retryFailed set to true?

Reply:
Fixed